tools/xmlschema.py

```diff
@@ -0,0 +1,462 @@
+                # # over sequences of unbounded choices with a permutation of bounded elements
+                # # inbetween them.
+                # container = ElementTree.Element(_to_qname("xs:choice"))
+                # for sequence in permutations(bounded_elements):
+                #     seq_container = ElementTree.Element(_to_qname("xs:sequence"))
+                #     seq_container.append(infinity_choice)
+                #     for pair in zip_longest(sequence, [], fillvalue=infinity_choice):
+                #         seq_container.extend(pair)
+                
+                #     container.append(seq_container)
+
+                # el.append(container)
+                # el.extend(attributes)
+
+                # above code appears to work, but the generated
+                # XSD is multiple GB in size. Use this for now.
+                seq_container = ElementTree.Element(_to_qname("xs:choice"))
+                seq_container.set("maxOccurs", "unbounded")
+                seq_container.extend(elements)
+                el.append(seq_container)
+                el.extend(attributes)
+
+        else:
+            el.extend(attributes)
+
+        return namespaces, el
+
+
+def setup_schema(used_ns: list, use_default_ns: bool = True) -> ElementTree.Element:
+    xsd_schema = ElementTree.Element(_to_qname("xs:schema"))
+    xsd_schema.set("version", "1.1")
+
+    if use_default_ns:
+        xsd_schema.set("xmlns", namespaces[source.stem])
+        xsd_schema.set("targetNamespace", namespaces[source.stem])
+
+    for name in set(used_ns):
+        el = ElementTree.Element(_to_qname("xs:import"))
+        el.set("namespace", namespaces[name])
+
+        if name == "types":
+            # types is a special class (not generated)
+            el.set("schemaLocation", f"./types.xsd")
+        else:
+            el.set("schemaLocation", f"./{name}Type.xsd")
+
+        xsd_schema.append(el)
+        xsd_schema.set(f"xmlns:{name}", namespaces[name])
+
+    return xsd_schema
+
+
+out_dir = Path(args.target)
+if not out_dir.exists():
+    out_dir.mkdir(exist_ok=True, parents=True)
+
+# write type file
+used_ns, element = ComplexType(root, root.attrib["name"] + "Type").to_subtree()
+xsd_schema = setup_schema(used_ns)
+xsd_schema.append(element)
+with open(out_dir / (source.stem + "Type.xsd"), "wb") as out_file:
+    ElementTree.ElementTree(xsd_schema).write(out_file, encoding="UTF-8", xml_declaration=True)
+
+# write element file
+file_name = source.stem
+tag_name = root.attrib["name"]
+used_ns, element = Element(root).to_basic()
+element.set("type", f"{file_name}:{tag_name}Type")
+used_ns.append(file_name)
+xsd_schema = setup_schema(used_ns, use_default_ns=False)
+xsd_schema.append(element)
+with open(out_dir / (source.stem + ".xsd"), "wb") as out_file:
+    ElementTree.ElementTree(xsd_schema).write(out_file, encoding="UTF-8", xml_declaration=True)
```

Comment:
The biggest problem I see is that the generated files are not readable by a human because the formatting is bad.
However, I also started to work on a Python version but gave it up after I found your PR because your approach is much better than mine.
To fix the problem with the unformatted output I used
```python
from xml.dom import minidom

with open(out_dir / (source.stem + "TypeF.xsd"), "w") as out_file:
    xmlstr = minidom.parseString(ElementTree.tostring(xsd_schema, xml_declaration=True, encoding='utf-8')).toprettyxml(indent="  ", encoding='utf-8').decode("utf-8")
    # replacing of < and > is needed for <![CDATA[text]]> but you don't use it
    #xmlstr = xmlstr.replace("&lt;", "<")
    #xmlstr = xmlstr.replace("&gt;", ">")
    # fixing text like 'The "ray", "gpu_ray", and "gps" types are...' after formatting
    xmlstr = xmlstr.replace("&quot;", '"')
    out_file.write(xmlstr)
```
to format the XML of Python's `xml.etree.ElementTree`. As mentioned in the formatting part your code doesn't use `<![CDATA[text]]>` while the current files created with the Ruby script [do](http://sdformat.org/schemas/sensor.xsd).

Reply:
I'm also not sure if `<xs:documentation xml:lang="en">` is required instead of `<xs:documentation>` since the language will always be English?

Reply:
> The biggest problem I see is that the generated files are not readable by a human because the formatting is bad.

Yes, we can definitely make the files more pretty. Once this gets picked up by one of the maintainers, I will have a closer look into that.

> As mentioned in the formatting part your code doesn't use <![CDATA[text]]> while the current files created with the Ruby script do.

Yes, `CDATA` is one way of escaping the comments. It has been a while since I've written the initial code, but I remember looking into it and deciding that we can also not have it here ... I have forgotten the reason for this, though, and I could have been wrong.

> I'm also not sure if <xs:documentation xml:lang="en"> is required instead of <xs:documentation> since the language will always be English?
 
Both options should work fine so I'm happy either way.